src/main/java/net/dries007/tfc/objects/recipes/UnmoldRecipe.java

```diff
@@ -26,21 +32,77 @@
 import net.dries007.tfc.objects.items.metal.ItemMetal;
 
 import static net.dries007.tfc.api.capability.heat.CapabilityItemHeat.ITEM_HEAT_CAPABILITY;
-import static net.dries007.tfc.api.util.TFCConstants.MOD_ID;
 import static net.minecraftforge.fluids.capability.CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY;
 
 @ParametersAreNonnullByDefault
-public class UnmoldRecipe extends IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
+public class UnmoldRecipe extends ShapelessOreRecipe
 {
-    private final ItemMold mold;
+    Metal.ItemType type;
+    float chance;
 
-    public UnmoldRecipe(ItemMold mold)
+    public UnmoldRecipe(ResourceLocation group, NonNullList<Ingredient> input, @Nonnull Metal.ItemType type, float chance)
     {
-        this.mold = mold;
+        super(group, input, ItemStack.EMPTY);
+        this.type = type;
+        this.chance = chance;
     }
 
     @Override
-    public boolean matches(InventoryCrafting inv, World worldIn)
+    @Nonnull
+    public NonNullList<ItemStack> getRemainingItems(final InventoryCrafting inv)
+    {
+        ItemStack moldStack = null;
+        for (int slot = 0; slot < inv.getSizeInventory(); slot++)
+        {
+            ItemStack stack = inv.getStackInSlot(slot);
+            if (!stack.isEmpty())
+            {
+                if (stack.getItem() instanceof ItemMold)
+                {
+                    ItemMold tmp = ((ItemMold) stack.getItem());
+                    if (tmp.type.equals(this.type) && moldStack == null)
+                    {
+                        moldStack = stack;
+                    }
+                    else
+                    {
+                        return super.getRemainingItems(inv);
+                    }
+                }
+                else
+                {
+                    return super.getRemainingItems(inv);
+                }
+            }
+        }
+        if (moldStack != null)
+        {
+            if (Constants.RNG.nextFloat() <= chance)
+            {
+                EntityPlayer player = ForgeHooks.getCraftingPlayer();
+                if (player != null)
+                {
+                    player.addItemStackToInventory(new ItemStack(moldStack.getItem()));
```

Comment:
As discussed in discord - if we make molds unstackable while filled, this be switched to just add to the return value i.e. how normal crafting recipes deal with container items?

Reply:
Because of the chance that it gets deleted this doesn't seem possible. There is no way I could find to update the crafting grid from the server, so I don't know how to get around the desync that rand causes

Reply:
You shouldn't be updating the crafting grid from anything *except* the server. The client shouldn't perform the crafting, the server should and sync any changes to the client

Reply:
Okay, I know, but when the server performs the craft the item doesn't appear in the grid, it stays invisible, how can I sync it to the client

Reply:
The same way containers sync their inventory normally - look into what `WorkbenchContainer` is doing and figure out why you have a desync bug.